# Import required libraries

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")

# Loading the dataset

In [4]:
df = pd.read_csv("C:\\Users\\chira\\OneDrive\\Desktop\\HousePricePrediction_ML_projects\\dataset\\Bengaluru_House_Data.csv")
df.head()

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00  
2      NaN       1440   2.0      3.0   62.00  
3  Soiewre       1521   3.0      1.0   95.00  
4      NaN       1200   2.0      1.0   51.00

# Shape of the dataset

In [6]:
df.shape

(13320, 9)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   area_type     13320 non-null  object 
 1   availability  13320 non-null  object 
 2   location      13319 non-null  object 
 3   size          13304 non-null  object 
 4   society       7818 non-null   object 
 5   total_sqft    13320 non-null  object 
 6   bath          13247 non-null  float64
 7   balcony       12711 non-null  float64
 8   price         13320 non-null  float64
dtypes: float64(3), object(6)
memory usage: 936.7+ KB


In [8]:
df.describe(include="all")

area_type   availability    location   size  society  \
count                  13320          13320       13319  13304     7818   
unique                     4             81        1305     31     2688   
top     Super built-up  Area  Ready To Move  Whitefield  2 BHK  GrrvaGr   
freq                    8790          10581         540   5199       80   
mean                     NaN            NaN         NaN    NaN      NaN   
std                      NaN            NaN         NaN    NaN      NaN   
min                      NaN            NaN         NaN    NaN      NaN   
25%                      NaN            NaN         NaN    NaN      NaN   
50%                      NaN            NaN         NaN    NaN      NaN   
75%                      NaN            NaN         NaN    NaN      NaN   
max                      NaN            NaN         NaN    NaN      NaN   

       total_sqft          bath       balcony         price  
count       13320  13247.000000  12711.000000  13320.000000  
unique       2117           NaN           NaN           NaN  
top          1200           NaN           NaN           NaN  
freq          843           NaN           NaN           NaN  
mean          NaN      2.692610      1.584376    112.565627  
std           NaN      1.341458      0.817263    148.971674  
min           NaN      1.000000      0.000000      8.000000  
25%           NaN      2.000000      1.000000     50.000000  
50%           NaN      2.000000      2.000000     72.000000  
75%           NaN      3.000000      2.000000    120.000000  
max           NaN     40.000000      3.000000   3600.000000

In [9]:
for column in df.columns:
    print(df[column].value_counts())
    print("*"*20)

area_type
Super built-up  Area    8790
Built-up  Area          2418
Plot  Area              2025
Carpet  Area              87
Name: count, dtype: int64
********************
availability
Ready To Move    10581
18-Dec             307
18-May             295
18-Apr             271
18-Aug             200
                 ...  
15-Aug               1
17-Jan               1
16-Nov               1
16-Jan               1
14-Jul               1
Name: count, Length: 81, dtype: int64
********************
location
Whitefield                        540
Sarjapur  Road                    399
Electronic City                   302
Kanakpura Road                    273
Thanisandra                       234
                                 ... 
Bapuji Layout                       1
1st Stage Radha Krishna Layout      1
BEML Layout 5th stage               1
singapura paradise                  1
Abshot Layout                       1
Name: count, Length: 1305, dtype: int64
********************
size
2 BHK    

# Dropping unnecessary columns

In [11]:
df2 = df.drop(['area_type', 'society', 'balcony', 'availability'], axis=1)
df2.head()

location       size total_sqft  bath   price
0  Electronic City Phase II      2 BHK       1056   2.0   39.07
1          Chikka Tirupathi  4 Bedroom       2600   5.0  120.00
2               Uttarahalli      3 BHK       1440   2.0   62.00
3        Lingadheeranahalli      3 BHK       1521   3.0   95.00
4                  Kothanur      2 BHK       1200   2.0   51.00

# Checking missing values

In [13]:
df2.isnull().sum()

location       1
size          16
total_sqft     0
bath          73
price          0
dtype: int64

In [14]:
df2.shape

(13320, 5)

# Handling missing values

In [16]:
df3 = df2.dropna() # Dropping null values
df3.isnull().sum()

location      0
size          0
total_sqft    0
bath          0
price         0
dtype: int64

# Data cleaning / transformation

In [18]:
df3['size'].unique()

array(['2 BHK', '4 Bedroom', '3 BHK', '4 BHK', '6 Bedroom', '3 Bedroom',
       '1 BHK', '1 RK', '1 Bedroom', '8 Bedroom', '2 Bedroom',
       '7 Bedroom', '5 BHK', '7 BHK', '6 BHK', '5 Bedroom', '11 BHK',
       '9 BHK', '9 Bedroom', '27 BHK', '10 Bedroom', '11 Bedroom',
       '10 BHK', '19 BHK', '16 BHK', '43 Bedroom', '14 BHK', '8 BHK',
       '12 Bedroom', '13 BHK', '18 Bedroom'], dtype=object)

* Here we clearly see that some mismatch reecords, like BHK and Bedroom. Therefore we need to fix them.

In [20]:
df3['bhk'] = df3['size'].apply(lambda x: int(x.split(' ')[0]))

In [21]:
df3['bhk'].unique()

array([ 2,  4,  3,  6,  1,  8,  7,  5, 11,  9, 27, 10, 19, 16, 43, 14, 12,
       13, 18], dtype=int64)

In [22]:
df3[df3['bhk']>20]

location        size total_sqft  bath  price  bhk
1718  2Electronic City Phase II      27 BHK       8000  27.0  230.0   27
4684                Munnekollal  43 Bedroom       2400  40.0  660.0   43

In [23]:
df3['total_sqft'].unique()

array(['1056', '2600', '1440', ..., '1133 - 1384', '774', '4689'],
      dtype=object)

* One can clearly see that there is a value "1133-1384" in the 'total-sqft' column. It's a range and we need to convert it into a single numeric value.

In [25]:
def convert_range(x):
    temp=x.split('-')
    if len(temp)==2:
        return (float(temp[0])+float(temp[1]))/2
    try:
        return float(x)
    except:
        return None

In [26]:
df3['total_sqft']=df3['total_sqft'].apply(convert_range)

In [27]:
df3.head()

location       size  total_sqft  bath   price  bhk
0  Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2
1          Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4
2               Uttarahalli      3 BHK      1440.0   2.0   62.00    3
3        Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3
4                  Kothanur      2 BHK      1200.0   2.0   51.00    2

* In the above code we make a user defined function "convert_range", temp = x.split('-'). It splits the string in to two numbers, like '1133-1384' to ['1133', '1384'], then if the length == 2 then it calculates the average of those two values.

In [29]:
df3['price_per_sqft']=df3['price']*100000/df3['total_sqft']
df3['price_per_sqft']

0         3699.810606
1         4615.384615
2         4305.555556
3         6245.890861
4         4250.000000
             ...     
13315     6689.834926
13316    11111.111111
13317     5258.545136
13318    10407.336319
13319     3090.909091
Name: price_per_sqft, Length: 13246, dtype: float64

In [30]:
df3.head()

location       size  total_sqft  bath   price  bhk  \
0  Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2   
1          Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4   
2               Uttarahalli      3 BHK      1440.0   2.0   62.00    3   
3        Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3   
4                  Kothanur      2 BHK      1200.0   2.0   51.00    2   

   price_per_sqft  
0     3699.810606  
1     4615.384615  
2     4305.555556  
3     6245.890861  
4     4250.000000

* created a new column i.e. "price_per_sqft"

In [32]:
df3['location'].nunique()

1304

In [33]:
df3['location']=df3['location'].apply(lambda x: x.strip()) #it removes leading or trailing whitespace characters(like spaces, tabs, or newlines) from x.

location_count=df3['location'].value_counts()

In [34]:
location_count

location
Whitefield                        535
Sarjapur  Road                    392
Electronic City                   304
Kanakpura Road                    266
Thanisandra                       236
                                 ... 
Vasantapura main road               1
Bapuji Layout                       1
1st Stage Radha Krishna Layout      1
BEML Layout 5th stage               1
Abshot Layout                       1
Name: count, Length: 1293, dtype: int64

In [35]:
# locations count less than 10
location_count_less_then_10=location_count[location_count<=10]
location_count_less_then_10

location
Naganathapura                     10
Sadashiva Nagar                   10
Nagappa Reddy Layout              10
BTM 1st Stage                     10
Sector 1 HSR Layout               10
                                  ..
Vasantapura main road              1
Bapuji Layout                      1
1st Stage Radha Krishna Layout     1
BEML Layout 5th stage              1
Abshot Layout                      1
Name: count, Length: 1052, dtype: int64

In [36]:
df3['location']=df3['location'].apply(lambda x: 'other' if x in location_count_less_then_10 else x)

In [37]:
df3['location'].value_counts()

location
other                 2881
Whitefield             535
Sarjapur  Road         392
Electronic City        304
Kanakpura Road         266
                      ... 
Nehru Nagar             11
Banjara Layout          11
LB Shastri Nagar        11
Pattandur Agrahara      11
Narayanapura            11
Name: count, Length: 242, dtype: int64

# Outlier detection and removing 

In [39]:
df4= df3[((df3["total_sqft"]/df3["bhk"])>=300)]
df4.shape

(12456, 7)

* Here we remove the rows, where the average area of each bedroom is less than 300 square feet.

In [41]:
df4.head()

location       size  total_sqft  bath   price  bhk  \
0  Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2   
1          Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4   
2               Uttarahalli      3 BHK      1440.0   2.0   62.00    3   
3        Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3   
4                  Kothanur      2 BHK      1200.0   2.0   51.00    2   

   price_per_sqft  
0     3699.810606  
1     4615.384615  
2     4305.555556  
3     6245.890861  
4     4250.000000

In [42]:
df4.describe()

total_sqft          bath         price           bhk  price_per_sqft
count  12456.000000  12456.000000  12456.000000  12456.000000    12456.000000
mean    1590.189927      2.562781    111.187960      2.649004     6308.502826
std     1260.404795      1.080275    152.203367      0.976046     4168.127339
min      300.000000      1.000000      9.000000      1.000000      267.829813
25%     1115.000000      2.000000     49.000000      2.000000     4210.526316
50%     1300.000000      2.000000     70.000000      3.000000     5294.117647
75%     1700.000000      3.000000    115.000000      3.000000     6916.666667
max    52272.000000     16.000000   3600.000000     16.000000   176470.588235

In [43]:
df4['price_per_sqft'].describe()

count     12456.000000
mean       6308.502826
std        4168.127339
min         267.829813
25%        4210.526316
50%        5294.117647
75%        6916.666667
max      176470.588235
Name: price_per_sqft, dtype: float64

* Here we got a minimum price_per_sqft of 267.82 and a maximum price_per_sqft of 176470.58 however a mean price_per_sqft of 6308.50 .

In [45]:
def remove_outliers_sqft(df):
    df_output = pd.DataFrame() # Initialize the empty DataFrame
    for key,subdf in df.groupby('location'):
        m = np.mean(subdf.price_per_sqft) # mean
        st= np.std(subdf.price_per_sqft) # Standard deviation

        gen_df = subdf[(subdf.price_per_sqft > (m-st)) & (subdf.price_per_sqft <= (m+st))]
        df_output = pd.concat([df_output,gen_df],ignore_index = True)
    return df_output
df5=remove_outliers_sqft(df4)
df5.shape

(10241, 7)

* here we used a user defining function(remove_outliers_sqft) for outlier treatment in price_per_sqft column.

In [47]:
df5.head()

location   size  total_sqft  bath  price  bhk  price_per_sqft
0  1st Block Jayanagar  4 BHK      2850.0   4.0  428.0    4    15017.543860
1  1st Block Jayanagar  3 BHK      1630.0   3.0  194.0    3    11901.840491
2  1st Block Jayanagar  3 BHK      1875.0   2.0  235.0    3    12533.333333
3  1st Block Jayanagar  3 BHK      1200.0   2.0  130.0    3    10833.333333
4  1st Block Jayanagar  2 BHK      1235.0   2.0  148.0    2    11983.805668

In [48]:
def bhk_outlier_remover(df):
    exclude_indices = np.array([])
    for location, location_df in df.groupby('location'):
        bhk_stats = {}
        for bhk, bhk_df in location_df.groupby('bhk'):
            bhk_stats[bhk] = {
                'mean': np.mean(bhk_df.price_per_sqft),
                'std': np.std(bhk_df.price_per_sqft),
                'count': bhk_df.shape[0]
            }

        for bhk, bhk_df in location_df.groupby('bhk'):
            stats = bhk_stats.get(bhk-1)
            if stats and stats['count']>5:
                exclude_indices = np.append(exclude_indices, bhk_df[bhk_df.price_per_sqft<(stats['mean'])].index.values)
    return df.drop(exclude_indices,axis='index')

In [49]:
df6=bhk_outlier_remover(df5)

In [50]:
df6.shape

(7329, 7)

In [51]:
df6['bath'].unique()

array([ 4.,  3.,  2.,  5.,  8.,  1.,  6.,  7.,  9., 12., 16., 13.])

In [52]:
df7=df6[df6['bath']<=df6['bhk']+1]
df7.head()

location   size  total_sqft  bath  price  bhk  price_per_sqft
0  1st Block Jayanagar  4 BHK      2850.0   4.0  428.0    4    15017.543860
1  1st Block Jayanagar  3 BHK      1630.0   3.0  194.0    3    11901.840491
2  1st Block Jayanagar  3 BHK      1875.0   2.0  235.0    3    12533.333333
3  1st Block Jayanagar  3 BHK      1200.0   2.0  130.0    3    10833.333333
4  1st Block Jayanagar  2 BHK      1235.0   2.0  148.0    2    11983.805668

In [53]:
df7.shape

(7251, 7)

* Filtering only those house records, where the number of bedrooms are equal or number of bedrooms is one less than the total number of bathrooms in that house.

In [55]:
df8 = df7.drop(['size','price_per_sqft'],axis=1)
df8.head()

location  total_sqft  bath  price  bhk
0  1st Block Jayanagar      2850.0   4.0  428.0    4
1  1st Block Jayanagar      1630.0   3.0  194.0    3
2  1st Block Jayanagar      1875.0   2.0  235.0    3
3  1st Block Jayanagar      1200.0   2.0  130.0    3
4  1st Block Jayanagar      1235.0   2.0  148.0    2

In [56]:
df8.shape

(7251, 5)

* Here we drop un-necessary columns like: 'size','price_per_sqft' from the dataframe.

# Converting categorical columns into numeric columns

In [59]:
df8['location'].nunique()

242

In [60]:
df9 = pd.get_dummies(df8['location'], dtype=int)
df9

1st Block Jayanagar  1st Phase JP Nagar  2nd Phase Judicial Layout  \
0                        1                   0                          0   
1                        1                   0                          0   
2                        1                   0                          0   
3                        1                   0                          0   
4                        1                   0                          0   
...                    ...                 ...                        ...   
10232                    0                   0                          0   
10233                    0                   0                          0   
10236                    0                   0                          0   
10237                    0                   0                          0   
10240                    0                   0                          0   

       2nd Stage Nagarbhavi  5th Block Hbr Layout  5th Phase JP Nagar  \
0                         0                     0                   0   
1                         0                     0                   0   
2                         0                     0                   0   
3                         0                     0                   0   
4                         0                     0                   0   
...                     ...                   ...                 ...   
10232                     0                     0                   0   
10233                     0                     0                   0   
10236                     0                     0                   0   
10237                     0                     0                   0   
10240                     0                     0                   0   

       6th Phase JP Nagar  7th Phase JP Nagar  8th Phase JP Nagar  \
0                       0                   0                   0   
1                       0                   0                   0   
2                       0                   0                   0   
3                       0                   0                   0   
4                       0                   0                   0   
...                   ...                 ...                 ...   
10232                   0                   0                   0   
10233                   0                   0                   0   
10236                   0                   0                   0   
10237                   0                   0                   0   
10240                   0                   0                   0   

       9th Phase JP Nagar  ...  Vishveshwarya Layout  Vishwapriya Layout  \
0                       0  ...                     0                   0   
1                       0  ...                     0                   0   
2                       0  ...                     0                   0   
3                       0  ...                     0                   0   
4                       0  ...                     0                   0   
...                   ...  ...                   ...                 ...   
10232                   0  ...                     0                   0   
10233                   0  ...                     0                   0   
10236                   0  ...                     0                   0   
10237                   0  ...                     0                   0   
10240                   0  ...                     0                   0   

       Vittasandra  Whitefield  Yelachenahalli  Yelahanka  Yelahanka New Town  \
0                0           0               0          0                   0   
1                0           0               0          0                   0   
2                0           0               0          0                   0   
3                0           0               0          0                   0   
4                0           0               0          0 

In [61]:
df10 = pd.concat([df8, df9.drop('other', axis=1)], axis=1)
df10.head()

location  total_sqft  bath  price  bhk  1st Block Jayanagar  \
0  1st Block Jayanagar      2850.0   4.0  428.0    4                    1   
1  1st Block Jayanagar      1630.0   3.0  194.0    3                    1   
2  1st Block Jayanagar      1875.0   2.0  235.0    3                    1   
3  1st Block Jayanagar      1200.0   2.0  130.0    3                    1   
4  1st Block Jayanagar      1235.0   2.0  148.0    2                    1   

   1st Phase JP Nagar  2nd Phase Judicial Layout  2nd Stage Nagarbhavi  \
0                   0                          0                     0   
1                   0                          0                     0   
2                   0                          0                     0   
3                   0                          0                     0   
4                   0                          0                     0   

   5th Block Hbr Layout  ...  Vijayanagar  Vishveshwarya Layout  \
0                     0  ...            0                     0   
1                     0  ...            0                     0   
2                     0  ...            0                     0   
3                     0  ...            0                     0   
4                     0  ...            0                     0   

   Vishwapriya Layout  Vittasandra  Whitefield  Yelachenahalli  Yelahanka  \
0                   0            0           0               0          0   
1                   0            0           0               0          0   
2                   0            0           0               0          0   
3                   0            0           0               0          0   
4                   0            0           0               0          0   

   Yelahanka New Town  Yelenahalli  Yeshwanthpur  
0                   0            0             0  
1                   0            0             0  
2                   0            0             0  
3                   0            0             0  
4                   0            0             0  

[5 rows x 246 columns]

In [62]:
df10.shape

(7251, 246)

In [63]:
cleaned_data=df10.drop(['location'],axis=1)
cleaned_data.shape

(7251, 245)

# Split the data to X and Y

In [65]:
X = cleaned_data.drop(['price'], axis=1)
y = cleaned_data['price']

In [66]:
X.shape

(7251, 244)

In [67]:
y.shape

(7251,)

# Divide into train test split

In [69]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Applying linear regression

In [71]:
from sklearn.linear_model import LinearRegression
lr_model = LinearRegression()
lr_model.fit(X_train, y_train)

LinearRegression()

In [72]:
lr_model.score(X_test, y_test)

0.8746160116850854

In [73]:
from sklearn.model_selection import GridSearchCV, ShuffleSplit
from sklearn.linear_model import LinearRegression, Lasso , Ridge
from sklearn.tree import DecisionTreeRegressor
from sklearn.preprocessing import StandardScaler
import pandas as pd

def find_best_model_using_gridsearchcv(X, y):
    # Scale the data if needed
    scaler = StandardScaler()
    X_scaled = scaler.fit_transform(X)

    algo = {
        'Linear_regression': {
            'model': LinearRegression(),
            'params': {
                'fit_intercept': [True, False]  # Removed 'normalize'
            }
        },
        'lasso': {
            'model': Lasso(),
            'params': {
                'alpha': [1,2],
                'selection': ['random', 'cyclic']
            }
        },
        'ridge': {
            'model': Ridge(),
            'params': {
                'alpha': [1,2]
            }
        },
        'Decision_tree':{
            'model': DecisionTreeRegressor(),
            'params': {
                'criterion': ['squared_error', 'friedman_mse'],  # Updated 'mse' to 'squared_error'
                'splitter': ['best', 'random'] 
            }
        }
        
    }
    scores = []
    cv = ShuffleSplit(n_splits=5, test_size=0.2, random_state=0)
    
    for algo_name, config in algo.items():
        gs = GridSearchCV(config['model'], config['params'], cv = cv, return_train_score=False)
        gs.fit(X_scaled, y)
        scores.append({
            'model': algo_name,
            'best_score': gs.best_score_,
            'best_params': gs.best_params_
        })
    return pd.DataFrame(scores, columns=['model', 'best_score', 'best_params'])
find_best_model_using_gridsearchcv(X, y)

model  best_score  \
0  Linear_regression    0.818354   
1              lasso    0.788752   
2              ridge    0.818442   
3      Decision_tree    0.722975   

                                         best_params  
0                            {'fit_intercept': True}  
1                {'alpha': 1, 'selection': 'cyclic'}  
2                                       {'alpha': 2}  
3  {'criterion': 'squared_error', 'splitter': 'be...

In [74]:
def predict_price(location, sqft, bath, bhk):
    # Get the index of the location column in the feature matrix
    loc_index = np.where(X.columns == location)[0][0]

    # Create a feature vector initialized with zeros
    x = np.zeros(len(X.columns))
    x[0] = sqft
    x[1] = bath
    x[2] = bhk

    # Set the location feature to 1 based on the location index
    if loc_index >= 0:
        x[loc_index] = 1

    # Convert the feature vector to a DataFrame with column names
    x = pd.DataFrame([x], columns=X.columns)  # This ensures the correct column names

    # Make the prediction using the model
    return lr_model.predict(x)[0]

In [75]:
X.columns

Index(['total_sqft', 'bath', 'bhk', '1st Block Jayanagar',
       '1st Phase JP Nagar', '2nd Phase Judicial Layout',
       '2nd Stage Nagarbhavi', '5th Block Hbr Layout', '5th Phase JP Nagar',
       '6th Phase JP Nagar',
       ...
       'Vijayanagar', 'Vishveshwarya Layout', 'Vishwapriya Layout',
       'Vittasandra', 'Whitefield', 'Yelachenahalli', 'Yelahanka',
       'Yelahanka New Town', 'Yelenahalli', 'Yeshwanthpur'],
      dtype='object', length=244)

* Predict_price is a user defined function created to predict the price of the house of bengaluru based on its location, sqft, no. of bathrooms and bhk.

In [77]:
predict_price('1st Phase JP Nagar', 1000, 4, 3)

91.61110615016443

In [78]:
predict_price('1st Block Jayanagar', 1200, 5, 5)

186.3976519062631

# Saving the file into a pickle format

In [80]:
import pickle
file_name = 'Bengaluru_House-Model.pkl'
pickle.dump(lr_model, open(file_name, 'wb'))

In [81]:
loaded_model = pickle.load(open('Bengaluru_House-Model.pkl', 'rb')) # load the model

In [82]:
import json
columns = {
    'data_columns': [col.lower() for col in X.columns]
}
with open("columns.json","w") as f:
    f.write(json.dumps(columns))